hyperspy/misc/utils.py

```diff
@@ -1228,21 +1194,22 @@ def process_function_blockwise(data,
     output_shape = chunk_nav_shape + tuple(output_signal_size)
     # Pre-allocating the output array
     output_array = np.empty(output_shape, dtype=dtype)
+    # Using deepcopy to avoid the function changing the original kwargs
```

Comment:
Why would the applied function changes the kwargs? If it does, then it is a bug in the applied function, not here, and I am not convinced that it is a concern of the `map`/`process_function_blockwise` function.
Deepcopying can have an expensive computing cost and it is 3 times here * the number of times process_function_blockwise is called, it can add very bad.

Reply:
The deepcopy is there to avoid some very strange and non-intuitive "bugs", where mutable objects are passed as arguments to the function used in `s.map`. For example a list or NumPy array. I added the deepcopy after some very strange behavior in a function.

A slightly contrived example of this:

```python
import numpy as np
import dask.array as da
import hyperspy.api as hs
def test_function(image, threshold):
    threshold[0] += 1
    mask = image > threshold[0]
    image *= mask
    return image
temp_data = np.arange(0, 20)
data = np.zeros((10, 10, 20))
for iy, ix in np.ndindex(data.shape[:-1]):
    data[iy, ix] = temp_data
data = da.from_array(data, chunks=(5, 5, 20))
s = hs.signals.Signal1D(data).as_lazy()
threshold = [10, 0, 0, 0]
s_out = s.map(test_function, threshold=threshold, inplace=False, lazy_result=False)
```

Gives the images. **With deepcopy**

![031_deepcopy_arg](https://user-images.githubusercontent.com/1690979/146034389-cf156f81-f64a-47c4-a311-b81456ada021.jpg)

**Without deepcopy**.

![031_not_deepcopy_arg](https://user-images.githubusercontent.com/1690979/146034415-ddf6f7b6-118c-4356-ba79-04aa64a1e59c.jpg)


In the last image here, where the deepcopy was not used, a different mask is applied for each position in each chunk. Leading to the different results.

----------

I do agree that this will have some impact on the performance when using `s.map`. However, for most arguments which are passed, this should be very low. For example if the arguments are a single `int`, `float` or `bool`. If iterating signals are passed as parameters, this will have more of an impact. However, the deepcopy will be limited to the specific navigation index which is being processed.

However, there is at least one situation where the deepcopy will have a pretty severe performance hit: if a very large NumPy array is passed as a non-iterating argument. Here, the whole array will be copied for every navigation position. But I think this an edge case.

Some tests for different dataset sizes:

**Large dataset, 64 x 64 x 512 x 512**

```python
from time import time
import numpy as np
import dask.array as da
import hyperspy.api as hs
def testfunction(image, mask):
    return image * mask
nav_x, nav_y = 64, 64
data = np.random.randint(0, 100, size=(nav_y, nav_x, 512, 512))
s = hs.signals.Signal2D(data)
data_mask = np.zeros((nav_y, nav_x, 512, 512), dtype=bool)
mask = np.zeros((512, 512), dtype=bool)
mask[0:20, 0:10] = True
for iy, ix in np.ndindex(nav_y, nav_x):
    data_mask[iy, ix, :, :] = mask
s_mask = hs.signals.Signal2D(data_mask)
###################################
s_out = s.map(testfunction, mask=s_mask, lazy_result=True, inplace=False)
###################################
t0 = time()
s_out.compute()
print(time() - t0)
# With deepcopy: 2.16 s, 2.15 s, 2.29 s, 2.29 s
# Without deepcopy: 2.06 s, 2.19 s, 2.2 s, 2.05 s
```

So the code with deepcopy is about 5% slower here.

**Small dataset, 2048 x 2048**

```python
from time import time
import numpy as np
import hyperspy.api as hs
def testfunction(image, mask):
    return image * mask
data = np.random.randint(0, 100, size=(2048, 2048))
s = hs.signals.Signal2D(data)
data_mask = np.zeros((2048, 2048), dtype=np.float64)
###################################
s_out = s.map(testfunction, mask=data_mask, lazy_result=True, inplace=False)
###################################
t0 = time()
s_out.compute()
print(time() - t0)
# With deepcopy: 0.101, 0.101, 0.101
# Without deepcopy: 0.1009, 0.101, 0.1009
```

Here, the difference is less than 1%.

-------------------

## Summary

I think we need some way of avoiding changes in mutable object "jumping" between navigation pixels. I guess ideally the functions people use shouldn't be changing the input parameters, but I don't think we can assume that. Thus, I fear that people will run into very strange and subtle behavior when using `s.map`.

There might be some other way than using `deepcopy`, but I'm not sure how this can be done. The worse performance is indeed a negative, but I feel that it is small enough to be worth it.

I guess one way of mitigating this, could be to include a `run_safe` parameter, where the deepcopy is not performed. This would default to `run_safe=True`, where people who want to maximize the performance could use `run_safe=False`.

Reply:
I agree that functions should not change their mutable arguments, and if they do that is a problem with that function. However, I worry that users will write their own custom made functions to be used with `map`, and since the "mutable object pitfall" is not well known among non-expert python programmers, I fear that users will get wrong results without even knowing it.

Also, without deepcopy, iterating signals which are passed to map can also be changed:

```python
import numpy as np
import hyperspy.api as hs
def test_function(image, iterating):
    iterating[0] = 2
    return image
s = hs.signals.Signal2D(np.zeros((1, 2, 20, 20)))
s_iter = hs.signals.Signal1D(np.zeros((1, 2, 2)))
s.map(test_function, iterating=s_iter)
s_iter
# array([[[2., 0.],
#         [2., 0.]]])
# All values here should be 0.
```

In addition, I'm not even sure if commonly used libraries such as scikit-image, scikit-learn, ... check for things like this? That the functions does not change its input arguments? I'm sure they try, but is this actually tested in the test suite?

------------

I think a compromise here could be an argument in `s.map`, `deepcopy_arguments`. Which by default is `True`, so that people who want to get the best performance can set it to `False`. I think most users would rather be sure that the results from `s.map` is correct, at a cost of a couple of percent of runtime.

Reply:
I added a `deepcopy_args` parameter to map, which toggles the deepcopying. It defaults to `True`.
I also added deepcopying for the input data as well, as I discovered that it is also affected by the same issue.

This also made it pretty easy to compare the performance for the deepcopy and non-deepcopy versions.

In short, the increase in runtime was about 0.86% to 1.9%, for the deepcopying. Ergo, doing the deepcopy makes the runtime about 1.5% longer.

-----------------
## Testing script

```python
from tqdm import trange
import time
import numpy as np
import hyperspy.api as hs

def mask_data(image, mask):
    image *= mask
    output = np.abs(np.fft.fftshift(np.fft.fft2(image)))
    return output[0, 0]

size = (10, 10, 500, 500)
s = hs.signals.Signal2D(np.random.random(size))
mask = np.zeros(size, dtype=bool)
mask[:, :, 200:, :] = True
s_mask = hs.signals.Signal2D(mask)

time_list_not_deepcopy = []
for i in trange(100):
    t0 = time.time()
    s_out = s.map(mask_data, mask=s_mask, inplace=False, deepcopy_args=False)
    time_list_not_deepcopy.append(time.time() - t0)
print("Signal size: {0}. Not deepcopy: {1}".format(s.axes_manager.shape, np.mean(time_list_not_deepcopy)))

time_list_deepcopy = []
for i in trange(100):
    t0 = time.time()
    s_out = s.map(mask_data, mask=s_mask, inplace=False, deepcopy_args=True)
    time_list_deepcopy.append(time.time() - t0)
print("Signal size: {0}. Deepcopy: {1}".format(s.axes_manager.shape, np.mean(time_list_deepcopy)))
```

----------------
## Results:

Signal size: (50, 50, 200, 200). Not deepcopy: 3.401603078842163
Signal size: (50, 50, 200, 200). Deepcopy: 3.4594222259521485
**1.7% slower**

Signal size: (2000, 2000). Not deepcopy: 0.5066886377334595
Signal size: (2000, 2000). Deepcopy: 0.5110295653343201
**0.85% slower**

Signal size: (4000, 4000). Not deepcopy: 2.219548602104187
Signal size: (4000, 4000). Deepcopy: 2.25101553440094
**1.4% slower**

Signal size: (10, 10, 500, 500). Not deepcopy: 0.9807310342788697
Signal size: (10, 10, 500, 500). Deepcopy: 0.9996297264099121
**1.9% slower**

Reply:
My opinion is that it simply needs to be explained to the docstring and it is not worth having an additional argument to the `map` method - the list of arguments is already long enough and I prefer to keep the API simple instead of adding argument to toggle workarounds.

When an user is changing arguments in the function itself, it sounds like an non-trivial usage of the `map` function and it would be fair to expect users to read the docstring correctly in this case or to be aware of this issue!